### PR TITLE
Do not write status file on Disable, Update. Create dummy status files during update if needed.

### PR DESCRIPTION
**Pull request by vivlingaiah**
- Do not write status file on Disable, Update. These two operations have been overwriting the status of previously executed Run command. So, not writing status files in these two operations will make status related to last execution (last enable ) available even after Update.
- Create dummy status files with status "success" corresponding to those .mrseq files that do not have status files during update. This is to temporarily correct the agent behavior where older agent versions delete the status file on disable during upgrade. After upgrade, there will be poll status timeouts without the status files for previously executed Run Commands. So, these dummy status files help fix that. Latest agent has fixed the issue. However, we need these dummy status files until old agent versions are being used in Production.
- Currently, timeouts are not an issue in production. This is a change in response to changes in agent where agent stopped removing the status files. Dummy status files are needed for older agents that are still in production, because older agents delete the status files in Disable operation that's part of upgrade.

### CODE REVIEW

File: internal/cmds/cmds.go [hunk 1]
```diff
@@ -334,21 +336,30 @@ func checkAndSaveSeqNum(ctx log.Logger, seq int, mrseqPath string) (shouldExit b
...
-		return err
+	_, statusFileCopyErr := copyFiles(ctx, ".status", constants.StatusFileDirectory)
+	if statusFileCopyErr != nil {
+		return statusFileCopyErr
 	}
+
+	// If status file corresponding to a .mrseq file does not exist, create a dummy status file to prevent poll status timeouts for already executed Run Commands after upgrade.
+	if mrseqFilesNameList != nil && mrseqFilesNameList.Len() > 0 {
+		createDummyStatusFilesErr := createDummyStatusFilesIfNeeded(ctx, mrseqFilesNameList)
+		if createDummyStatusFilesErr != nil {
```
Comment (jscalev) [inline internal/cmds/cmds.go:353]: I'm wondering if this should be best effort. What if there's a race condition where an executing RC tries to write the file? Best effort (perhaps with a trace) seems like a safe approach.

File: internal/cmds/cmds.go [hunk 2]
```diff
@@ -422,16 +434,89 @@ func copyFiles(ctx log.Logger, fileExtensionSuffix string, extensionSubdirectory
...
+		if err != nil {
+			ctx.Log("error", fmt.Sprintf("Reading mrseq (Most Recently executed Sequence number) from file '%s' failed with error '%s'", mrSeqFileFullPath, err.Error()))
+			return err
+		}
+
+		var mrseqNumber int
+		if content != nil {
+			mrseqNumberString := string(content)
+			mrseqNum, errr := strconv.Atoi(mrseqNumberString)
+			if errr != nil {
```
Comment (jscalev) [inline internal/cmds/cmds.go:482]: Why not reuse err?

File: internal/cmds/cmds.go [hunk 3]
```diff
@@ -422,16 +434,89 @@ func copyFiles(ctx log.Logger, fileExtensionSuffix string, extensionSubdirectory
...
+	statusFileDirectoryPath := filepath.Join(newExtensionDirectory, constants.StatusFileDirectory)
+
+	var mrSeqFileName string
+	var mrSeqFileFullPath string
+	var extensionName string
+	var mrSeqFileExtensionIndex int
+	var statusFileName string
+	var statusFilePath string
+	for mreSeqFileNameElement := mrseqFilesNameList.Front(); mreSeqFileNameElement != nil; mreSeqFileNameElement = mreSeqFileNameElement.Next() {
+		mrSeqFileName = (mreSeqFileNameElement.Value).(string)
```
Comment (jscalev) [inline internal/cmds/cmds.go:468]: should this be best effort? If we fail to read one .mrseq, should we just move past that one and try to write dummy files for the others?

File: internal/cmds/cmds.go [hunk 4]
```diff
@@ -422,16 +434,89 @@ func copyFiles(ctx log.Logger, fileExtensionSuffix string, extensionSubdirectory
...
+		}
+		extensionName = mrSeqFileName[0:mrSeqFileExtensionIndex]
+
+		// Determine status file name and status file path
+		statusFileName = fmt.Sprintf("%s.%d.status", extensionName, mrseqNumber)
+		statusFilePath = filepath.Join(statusFileDirectoryPath, statusFileName)
+
+		// If status file path does not exist, create a dummy status file to prevent poll status timeouts for already executed Run Commands after upgrade.
+		if !handlersettings.DoesFileExist(statusFilePath) {
+			statusReport := types.NewStatusReport(types.StatusSuccess, "Enable", "The script has been executed. However, the real execution state, output, error are unknown.")
```
Comment (jscalev) [inline internal/cmds/cmds.go:508]: Probably should be more specific and state that the previous status file was deleted and we've replaced it but without the execution state, output, and error.

File: internal/cmds/cmds.go (hunk 2 above)
Comment (vivlingaiah) [reply internal/cmds/cmds.go:482]: done

File: internal/cmds/cmds.go (hunk 4 above)
Comment (vivlingaiah) [reply internal/cmds/cmds.go:508]: These are the only states CRP understands. So, using "Warning" state now. CRP displays "warning" in instanceView, but it treats as ProivisioningState = Succeeded at the RC ARM resource level.

    // State of an extension as reported by the extension handler.
    public enum ExtensionStates
    {
        Transitioning,
        Error,
        Success,
        Warning,
    }

File: internal/cmds/cmds.go (hunk 3 above)
Comment (vivlingaiah) [reply internal/cmds/cmds.go:468]: Yes, agree

File: internal/cmds/cmds.go (hunk 1 above)
Comment (vivlingaiah) [reply internal/cmds/cmds.go:353]: Yes, I agree it should be best effort. Changed accordingly. Added traces inside createDummyStatusFilesNeeded()
Race condition - I do not expect a race condition to happen. createDummyStatusFilesNeeded() is called only during Update operation. Enable operations that can write to status file only comes later once Update has completed.

**Review by narrieta — APPROVED** [review]

**Review by jscalev — APPROVED** [review]

**Pull request merged**